### PR TITLE
[patreon] fix some vimeo embed downloads

### PR DESCRIPTION
**Pull request by johnvictorfs**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

~~This change prioritizes downloading from m3u8 instead of vimeo embeds (or embeds in general), which in some cases may no longer exist and only the m3u8 video attachment does.~~

This change adds checks to verify if external embeds of Patreon posts are still valid before trying to download them (since they may have been deleted from vimeo/dropbox/etc. but still be attached to the Patreon post), and if it is invalid then it tries other options

Example, using `python -m yt_dlp https://www.patreon.com/posts/hunter-x-hunter-34007913` (this video is public, so no need for cookies/account to test with)

```sh
# Before change, finds a vimeo embed (that no longer exists) and fails to download it
python -m yt_dlp https://www.patreon.com/posts/hunter-x-hunter-34007913
[Patreon] Extracting URL: https://www.patreon.com/posts/hunter-x-hunter-34007913
[Patreon] 34007913: Downloading API JSON
[vimeo] Extracting URL: https://player.vimeo.com/video/391366663?app_id=122963#__youtubedl_smuggle=%7B%22referer%22%3A+%2...%2Fpatreon.com%22%7D
[vimeo] 391366663: Downloading webpage
ERROR: [vimeo] 391366663: Unable to download webpage: HTTP Error 404: Not Found (caused by <HTTPError 404: Not Found>)
```

```sh
# After change: checks for vimeo embed, fails, then tries to download the m3u8 file attachment
python -m yt_dlp https://www.patreon.com/posts/hunter-x-hunter-34007913
[Patreon] Extracting URL: https://www.patreon.com/posts/hunter-x-hunter-34007913
[Patreon] 34007913: Downloading API JSON
[Patreon] 34007913: Downloading webpage
WARNING: [Patreon] Unable to download webpage: HTTP Error 405: Method Not Allowed
[Patreon] 34007913: Downloading webpage
WARNING: [Patreon] Unable to download webpage: HTTP Error 404: Not Found
[Patreon] 34007913: Downloading m3u8 information
[info] 34007913: Downloading 1 format(s): 4712
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 284
[download] Destination: Hunter x Hunter ｜ Kurapika DESTROYS Uvogin!!! [34007913].mp4
[download]  10.2% of ~ 721.42MiB at    2.61MiB/s ETA 03:26 (frag 28/284)
```

**Possibly** fixes https://github.com/yt-dlp/yt-dlp/issues/8702 (don't have access to the post in the example to test but seems pretty similar to my test case)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by pukkandan** [comment]
> which in some cases may no longer exist and only the m3u8 video attachment does.

Do you know why this happens? Uploader deletes vimeo video? Can the reverse scenario happen?

**Comment by johnvictorfs** [comment]
> > which in some cases may no longer exist and only the m3u8 video attachment does.
> 
> Do you know why this happens? Uploader deletes vimeo video? Can the reverse scenario happen?
> 

@pukkandan Not 100% sure but yeah I believe the uploader deleting just the Vimeo video could be it (or Vimeo themselves). The reverse scenario seems really unlikely to me (although I am not a creator on Patreon so I can't test that) since the removal of a Patreon embed would be done on Patreon itself, I would image it would be removed from the post correctly if that was the case.

> Also, if both work, we should prioritize the video embed, correct?

But yes, in that scenario prioritizing the embed would make more sense, I could change it so it would still try the embed first but if it fails try other options, I will look into that a bit later, thanks!

**Comment by pukkandan** [comment]
We can make a HeadRequest to the Vimeo URL and fall back to m3u8 if that fails

**Comment by johnvictorfs** [comment]
> We can make a HeadRequest to the Vimeo URL and fall back to m3u8 if that fails

@pukkandan Changed the code to use this approach, is it correct?

Don't have any public posts I could share that fall into this example (having both a working embed and a file attachment), but seems to be working with some private posts I have access to.

image

I also haven't found any working vimeo embeds on Patreon to test with (just dropbox) but it worked with a hard-coded vimeo url temporarily in the code as if it was attached to the post:

image

My original test (with a public post) also works fine

```sh
python -m yt_dlp https://www.patreon.com/posts/hunter-x-hunter-34007913
[Patreon] Extracting URL: https://www.patreon.com/posts/hunter-x-hunter-34007913
[Patreon] 34007913: Downloading API JSON
[Patreon] 34007913: Downloading webpage
WARNING: [Patreon] Unable to download webpage: HTTP Error 405: Method Not Allowed
[Patreon] 34007913: Downloading webpage
WARNING: [Patreon] Unable to download webpage: HTTP Error 404: Not Found
[Patreon] 34007913: Downloading m3u8 information
[info] 34007913: Downloading 1 format(s): 4712
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 284
[download] Destination: Hunter x Hunter ｜ Kurapika DESTROYS Uvogin!!! [34007913].mp4
[download]  10.2% of ~ 721.42MiB at    2.61MiB/s ETA 03:26 (frag 28/284)
```

**Comment by bashonly** [comment]
> HTTP Error 405: Method Not Allowed

this means a HEAD request won't work

**Comment by johnvictorfs** [comment]
> > HTTP Error 405: Method Not Allowed
> 
> this means a HEAD request won't work. IMO just use GET for both

@bashonly Yeah you're right, I missed that, only tested vimeo.com links and not player.vimeo.com, changing both to GET requests, thanks.

File: yt_dlp/extractor/patreon.py [hunk 1]
```diff
@@ -250,20 +250,27 @@ def _real_extract(self, url):
...
+                vimeo_embed_url = VimeoIE._smuggle_referrer(v_url, 'https://patreon.com')
+                response = self._request_webpage(vimeo_embed_url, video_id, fatal=False)
+
+                if response is not False:
+                    return {
+                        **info,
+                        '_type': 'url_transparent',
+                        'url': vimeo_embed_url,
+                        'ie_key': 'Vimeo',
+                    }
```
Comment (bashonly) [inline yt_dlp/extractor/patreon.py:262]:
```suggestion
                v_url = VimeoIE._smuggle_referrer(v_url, 'https://patreon.com')
                if self._request_webpage(v_url, video_id, 'Checking Vimeo embed URL', fatal=False, errnote=False):
                    return {
                        **info,
                        '_type': 'url_transparent',
                        'url': vimeo_embed_url,
                        'ie_key': 'Vimeo',
                    }
```

File: yt_dlp/extractor/patreon.py [hunk 2]
```diff
@@ -250,20 +250,27 @@ def _real_extract(self, url):
...
         embed_url = try_get(attributes, lambda x: x['embed']['url'])
         if embed_url:
-            return {
-                **info,
-                '_type': 'url',
-                'url': embed_url,
-            }
+            response = self._request_webpage(embed_url, video_id, fatal=False)
+
+            if response is not False:
```
Comment (bashonly) [inline yt_dlp/extractor/patreon.py:268]:
```suggestion
                if self._request_webpage(embed_url, video_id, 'Checking embed URL', fatal=False, errnote=False):
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/patreon.py (hunk 1 above)
Comment (johnvictorfs) [reply yt_dlp/extractor/patreon.py]: Updated

File: yt_dlp/extractor/patreon.py [hunk 3]
```diff
@@ -250,20 +250,24 @@ def _real_extract(self, url):
...
-                }
+                v_url = VimeoIE._smuggle_referrer(v_url, 'https://patreon.com')
+
+                if self._request_webpage(v_url, video_id, 'Checking Vimeo embed URL', fatal=False, errnote=False):
+                    return {
+                        **info,
+                        '_type': 'url_transparent',
+                        'url': v_url,
+                        'ie_key': 'Vimeo',
+                    }
```
Comment (Grub4K) [inline yt_dlp/extractor/patreon.py:261]: maybe this should be
```suggestion
                    return self.url_result(v_url, 'Vimeo', **info, url_transparent=True)
```

File: yt_dlp/extractor/patreon.py [hunk 4]
```diff
@@ -250,20 +250,24 @@ def _real_extract(self, url):
...
-                **info,
-                '_type': 'url',
-                'url': embed_url,
-            }
+            if self._request_webpage(embed_url, video_id, 'Checking embed URL', fatal=False, errnote=False):
+                return {
+                    **info,
+                    '_type': 'url',
+                    'url': embed_url,
+                }
```
Comment (Grub4K) [inline yt_dlp/extractor/patreon.py:270]: same
```suggestion
                return self.url_result(embed_url, **info)
```

**Review by Grub4K — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

**Pull request merged**